### PR TITLE
Develop module first version

### PR DESCRIPTION
**Pull request by ramin-fs**
https://trello.com/c/NJusBZVq/103-pod-checkmodulespodcoverage

### CODE REVIEW

File: Changes [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{{$NEXT}}
```
Comment (tom-binary) [inline Changes]: This is replaced by the version information

File: Changes [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+
+Revision history for test-pod-coverage-syntax
```
Comment (tom-binary) [inline Changes:1]: This should be at the top of the file, *above* the `{{$NEXT}}` template directive

File: Changes [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+
+0.01      2020-06-26 00:00:00+00:00 UTC
```
Comment (tom-binary) [inline Changes]: This needs to be removed, it has not been released yet.

File: Changes [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+        First version, released on an unsuspecting world.
```
Comment (tom-binary) [inline Changes]:
```suggestion
        Initial version.
```

File: lib/Test/Pod/CoverageChange.pm [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+=head1 SYNOPSIS
+
+It checks all files that placed under the `lib` folder against their POD coverage to see if all existing functions have POD or not?
+It checks all files that placed under the `lib` folder against their POD syntax to see if they have a valid POD syntax or not.
+
+=head1 DESCRIPTION
+
+Prints the percentage of POD coverage in **TODO** test format for currently existing packages.
+Prints B<not ok> (with respective message) if our latest changes increased/decreased numbers of naked functions in currently existing packages.
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: as mentioned in Trello, should be `C<>`

File: lib/Test/Pod/CoverageChange.pm [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+Pod coverage calculator test file.
+
+=head1 SYNOPSIS
+
+It checks all files that placed under the `lib` folder against their POD coverage to see if all existing functions have POD or not?
+It checks all files that placed under the `lib` folder against their POD syntax to see if they have a valid POD syntax or not.
+
+=head1 DESCRIPTION
+
+Prints the percentage of POD coverage in **TODO** test format for currently existing packages.
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: Don't use markdown in POD - why is this marked as "TODO", what should be here?

File: lib/Test/Pod/CoverageChange.pm [hunk 7]
```diff
@@ -0,0 +1,188 @@
+package Test::Pod::CoverageChange;
+# ABSTRACT: check perl modules against their pod coverage
+
+=pod
+
+=head1 NAME
+
+Pod coverage calculator test file.
+
+=head1 SYNOPSIS
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm:18]: Synopsis should contain verbatim code showing an example of how to use the module.

File: lib/Test/Pod/CoverageChange.pm [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+=head1 DESCRIPTION
+
+Prints the percentage of POD coverage in **TODO** test format for currently existing packages.
+Prints B<not ok> (with respective message) if our latest changes increased/decreased numbers of naked functions in currently existing packages.
+Prints B<not ok> if currently existing packages have 100% coverage unless you remove the package from B<%naked_packages> variable.
+Prints a proper message for the newly added packages.
+
+prints B<ok> for the files that have no POD syntax error.
+prints B<not ok- There is no POD in the file> if the file has no POD at all. I put this into a TODO test so CircleCI's tests will pass.
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: There is no CircleCI - this is a public module.

File: lib/Test/Pod/CoverageChange.pm [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+prints B<not ok- The number of errors in the POD structure> if the file has any error. It causes CircleCI's tests to fail.
+
+=cut
+use strict;
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: POD is paragraph-based, needs blank lines before+after every `=` directive.

File: lib/Test/Pod/CoverageChange.pm [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+use Module::Path 'module_path';
+
+our $VERSION='0.01';
+
+use constant {
+  POD_SYNTAX_IS_OK => 0,
+  FILE_HAS_NO_POD => -1,
+};
+
+our @ISA = qw(Exporter);
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: never use `our @ISA`, it'd be `use parent` and we shouldn't be doing this anyway:

```suggestion
use Exporter qw(import export_to_level);
```

File: lib/Test/Pod/CoverageChange.pm [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+=head2 check
+
+Todo: Update this pod
+
+Check all modules under the given directory against POD coverage and POD syntax
+
+=over 4
+
+=item * C<directory> An arrayref of directories to check pod against
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]:
```suggestion
=item * C<directory> - arrayref of directories to check pod against
```

File: lib/Test/Pod/CoverageChange.pm [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    my $naked_packages = shift // {};
+    my $ignored_packages = shift // [];
+
+    $path = [$path] unless ref $path eq 'ARRAY';
+    $ignored_packages = [$ignored_packages] unless ref $ignored_packages eq 'ARRAY';
+
+    check_pod_coverage($path, $naked_packages, $ignored_packages);
+    check_pod_syntax($path, $ignored_packages);
+
+    Test::Pod::CoverageChange->export_to_level(1, @_);
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: Why is this here?

File: lib/Test/Pod/CoverageChange.pm [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    my $directories = shift;
+    my $allowed_naked_packages = shift;
+    my $ignored_packages = shift;
+
+    check_existing_naked_packages($allowed_naked_packages, $ignored_packages) if keys %$allowed_naked_packages;
+
+    # Check for newly added packages PODs
+    my @ignored_packages = (keys %$allowed_naked_packages, @$ignored_packages);
+    foreach my $package (Test::Pod::Coverage::all_modules(@$directories)) {
+        next if @ignored_packages && (grep(/^$package$/, @ignored_packages));
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: If this is supposed to support regex and do partial matching, then you wouldn't want the `^$` anchors. If it's a string match, then use a string match...

```suggestion
        next if @ignored_packages && (any { $_ eq $package } @ignored_packages));
```

or preferably:

```
my %ignored;
@ignored{@ignored_packages} = ();
```

so that you can make this a simple `next if exists $ignored{$package}`

File: lib/Test/Pod/CoverageChange.pm [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    my $ignored_packages = shift;
+
+    my @ignored_packages_full_path = ();
+    for (@$ignored_packages) {
+        my $file_path = module_path($_);
+        push(@ignored_packages_full_path, $file_path) if defined $file_path;
+    }
+
+    my @files_path = File::Find::Rule->file()
+                                     ->name( '*.p[m|l]' )
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: generally alignment should match the level of indentation (1 for the sub, another for the incomplete statement), otherwise you end up with arbitrary spacing based just on function/package names

File: t/correct_pod.t [hunk 15]
```diff
@@ -0,0 +1,21 @@
+use warnings;
```
Comment (tom-binary) [inline t/correct_pod.t]: strict first, then warnings - for some reason all these have them swapped?

File: cpanfile [hunk 16]
```diff
@@ -0,0 +1,27 @@
+requires "Exporter" => "0";
+requires "Moo" => "1.006000";
```
Comment (tom-binary) [inline cpanfile]: Where are we using this?

File: cpanfile [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+requires "Pod::Checker" => "0";
+requires "Pod::Coverage" => "0";
+requires "Test::More" => "0";
+requires "Test::Pod::Coverage" => "0";
+requires "WebService::Client" => "0.0001";
```
Comment (tom-binary) [inline cpanfile]: This also seems unused?

File: cpanfile [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+requires "base" => "0";
```
Comment (tom-binary) [inline cpanfile]: not used

File: cpanfile [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+requires "constant" => "0";
+requires "strict" => "0";
```
Comment (tom-binary) [inline cpanfile]: core modules, `strict` and `warnings` are never required in `cpanfile`

File: t/correct_pod.t (hunk 15 above)
Comment (ramin-fs) [reply t/correct_pod.t]: My code editor changes the order. I'll fix it.

File: lib/Test/Pod/CoverageChange.pm (hunk 5 above)
Comment (ramin-fs) [reply lib/Test/Pod/CoverageChange.pm]: I'm going to update all the pods based on the Trello comment. I'll notify you after it is done.

File: cpanfile [hunk 20]
```diff
@@ -0,0 +1,22 @@
+requires "Exporter" => "0";
```
Comment (tom-binary) [inline cpanfile]: see our other `cpanfile` examples - please don't use `=>` here!

File: cpanfile [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+requires "Pod::Checker" => "0";
+requires "Pod::Coverage" => "0";
+requires "Test::More" => "0";
+requires "Test::Pod::Coverage" => "0";
+requires "constant" => "0";
+
+on 'test' => sub {
+  requires "File::Spec" => "0";
```
Comment (tom-binary) [inline cpanfile]: 4-space indent

File: lib/Test/Pod/CoverageChange.pm [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+
+=head2 check_pod_coverage
+
+Checks POD coverage for all the modules that exists under a given directory.
+Passes the $allowed_naked_packages to the L<Test::Pod::CoverageChange::check_allowed_naked_packages>
+Ignores the packages in the C<$ignored_packages> parameter
+
+=over 4
+
+=item C<$directories> Arrayref|String - I works on all the modules that lies under this given directory(ies).
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: again, please look at other examples - we don't use this style anywhere else in the code

```suggestion
=item C<$directories> - directories to check recursively, supports string or arrayref
```

File: cpanfile [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+requires "Pod::Coverage";
+requires "Test::More";
+requires "Test::Pod::Coverage";
+requires "constant";
+
+on 'test' => sub {
+    requires "File::Spec";
+    requires "IO::Handle";
+    requires "IPC::Open3";
+    requires "Test::More";
```
Comment (zakame) [inline cpanfile]: Because your tests use `subtest`, this needs 
```suggestion
    requires "Test::More", "0.98";
```

File: cpanfile [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+    requires "perl", ">= 5.006";
```
Comment (zakame) [inline cpanfile]: We probably don't need to require such an old perl, and more to the point, probably should go as a `runtime_requires`, not on `test_requires`.

File: dist.ini [hunk 25]
```diff
@@ -0,0 +1,72 @@
+name    = perl-Test-Pod-CoverageChange
+author  = Binary Group Services Ltd <BINARY@cpan.org>
+license = Perl_5
+copyright_holder = Binary Group Services Ltd
+copyright_year   = 2020
+main_module= lib/Test/Pod/CoverageChange.pm
+
+[Git::GatherDir]
```
Comment (zakame) [inline dist.ini]: Maybe https://metacpan.org/pod/Dist::Zilla::PluginBundle::Author::DERIV already provides for the rest of this?

File: t/no_pod.t [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+
+# Initializing variables
+my $test_module = "t::Nopod";
+my $test_module_path = 't/Nopod.pm';
+my $main_module_path = module_path('Test::Pod::CoverageChange');
+
+subtest 'Module with no pod, unexpected' => sub {
+    test_out("not ok 1 - Pod coverage on $test_module");
+    test_out("not ok 2 # TODO There is no POD in the file $test_module_path.");
+    test_out("#   Failed (TODO) test at $main_module_path line 175.");
```
Comment (ramin-fs) [inline t/no_pod.t]: @zakame can you help me with this, please? I want to get rid of this filename and line number but I've couldn't handle it. Do you have any idea to do that?

File: cpanfile (hunk 24 above)
Comment (ramin-fs) [reply cpanfile]: I completely removed it. I copied it from another module and I didn't know if it better to remove it or keep it. thanks.

File: t/no_pod.t (hunk 26 above)
Comment (zakame) [reply t/no_pod.t]: So if I read this right, this TODO fail is raised via https://github.com/binary-com/perl-Test-Pod-CoverageChange/blob/9026e1d4ea3aa1cde400c73b699e3e7bbd1757c8/lib/Test/Pod/CoverageChange.pm#L175-L180 and what you're trying to do is to raise this TODO fail from the context of this test script, not from this `CoverageChange` module, a la Carp.  Per https://metacpan.org/pod/Test::More#Extending-and-Embedding-Test::More, you probably might need to check via the underlying `Test::Builder` instead and use https://metacpan.org/pod/Test::Builder#todo_start and https://metacpan.org/pod/Test::Builder#todo_skip ...

Comment (ramin-fs) [reply t/no_pod.t]: @zakame I change these parts. Thanks for your great comment.

File: dist.ini (hunk 25 above)
Comment (ramin-fs) [reply dist.ini]: you mean I replace all the content below this line with `[Author::DERIV]`? If the answer is yes I change it a few minutes ago but I'm not able to install the CPAN module for this plugin. Can you please help me with that?

File: cpanfile [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+requires "Pod::Coverage";
+requires "Test::More";
+requires "Test::Pod::Coverage";
+requires "constant";
+
+on 'test' => sub {
+    requires "File::Spec";
+    requires "IO::Handle";
+    requires "IPC::Open3";
+    requires "Test::More", "0.98";
```
Comment (tom-binary) [inline cpanfile]:
```suggestion
    requires "Test::More", ">= 0.98";
```

File: cpanfile [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+};
+
+on 'configure' => sub {
+    requires "ExtUtils::MakeMaker";
+};
+
+on 'develop' => sub {
+    requires "Test::Pod", "1.41";
```
Comment (tom-binary) [inline cpanfile]:
```suggestion
    requires "Test::Pod", ">= 1.41";
```

File: cpanfile [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+requires "constant";
+
+on 'test' => sub {
+    requires "File::Spec";
+    requires "IO::Handle";
+    requires "IPC::Open3";
+    requires "Test::More", "0.98";
+};
+
+on 'configure' => sub {
```
Comment (tom-binary) [inline cpanfile]: this should be removed entirely

File: lib/Test/Pod/CoverageChange.pm [hunk 30]
```diff
@@ -0,0 +1,246 @@
+package Test::Pod::CoverageChange;
+# ABSTRACT: check perl modules against their pod coverage
+
+=pod
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: not required, should be removed

File: lib/Test/Pod/CoverageChange.pm [hunk 31]
```diff
@@ -0,0 +1,246 @@
+package Test::Pod::CoverageChange;
+# ABSTRACT: check perl modules against their pod coverage
+
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm:6]: strict, warnings and POD encoding should always be at the top of the file. Would also need the $VERSION definition.

```
use strict;
use warnings;

our $VERSION = '0.001';
# AUTHORITY

use utf8;

=encoding utf8
```

File: lib/Test/Pod/CoverageChange.pm [hunk 32]
```diff
@@ -0,0 +1,246 @@
...
+=head1 NAME
+
+Pod coverage calculator test file.
+
+=head1 SYNOPSIS
+
+    use Test::Pod::CoverageChange qw(check);
+
+    Test::Pod::CoverageChange::check('lib', {
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: if this is being imported, then it wouldn't need to be fully qualified - but I'd suggest renaming, `check` is way too generic if there are other tests in here. Consider `pod_coverage_ok` or similar.

File: lib/Test/Pod/CoverageChange.pm [hunk 33]
```diff
@@ -0,0 +1,246 @@
...
+POD coverage to see if all existing subs have POD and also
+POD syntax to see if is there an POD syntax error or not?
+
+Prints the percentage of POD coverage in B<TODO> test format for the packages that we allowed to have naked subs.
+Prints an error message if our latest changes increased/decreased numbers of naked sub for the packages that we allowed to have naked sub.
+Prints an error message if a naked allowed package has 100% POD coverage. (We should remove it from the C<%naked_packages> variable list.)
+Ignores to check every package that we pass as C<@ignored_package>
+Prints a proper message for the newly added packages.
+
+Prints C<ok> for the files that have no POD syntax error.
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: This is inaccurate - this shouldn't be *printing* anything, it's running tests. `prove -v` happens to render this as raw TAP output, but there are many other test harnesses and formatters.

File: lib/Test/Pod/CoverageChange.pm [hunk 34]
```diff
@@ -0,0 +1,246 @@
...
+Prints C<not ok- There is no POD in the file> if the file has no POD at all. I will consider as a pass test since it is in a todo section.
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]:
```suggestion
If the file has no POD at all, it will generate a failing TODO test.
```

File: lib/Test/Pod/CoverageChange.pm [hunk 35]
```diff
@@ -0,0 +1,246 @@
...
+
+use strict;
+use warnings;
+
+use Test::More;
+use Pod::Checker;
+use Pod::Coverage;
+use File::Find::Rule;
+use Test::Pod::Coverage;
+use Module::Path 'module_path';
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: please think about consistency when you're writing code - the line below is using `qw()` for the list, but for some reason you have a different quoting style here...

```suggestion
use Module::Path qw(module_path);
```

File: lib/Test/Pod/CoverageChange.pm [hunk 36]
```diff
@@ -0,0 +1,246 @@
...
+use List::Util qw(any);
+
+our $VERSION = '0.01';
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: Each part of a version in Perl has 3 digits, so I would recommend against 2-digit versions like this. Also, should be at the top of the file.

File: lib/Test/Pod/CoverageChange.pm [hunk 37]
```diff
@@ -0,0 +1,246 @@
...
+use constant {
+    POD_SYNTAX_IS_OK => 0,
+    FILE_HAS_NO_POD  => -1,
+};
+
+use Exporter qw(import export_to_level);
+our @EXPORT_OK = qw(check);
+
+# Set caller test file name
+my $caller_test_file = (caller())[1];
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: This should be in the test function, not at the top level.

File: lib/Test/Pod/CoverageChange.pm [hunk 38]
```diff
@@ -0,0 +1,246 @@
...
+
+# Create a test builder object
+my $Test_Builder = Test::More->builder;
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: This should also be in the test function or at least created on demand.

File: lib/Test/Pod/CoverageChange.pm [hunk 39]
```diff
@@ -0,0 +1,246 @@
...
+
+=pod
+
+=head2 check
+
+Check all modules under a given directory against POD coverage and POD syntax
+
+=over 4
+
+=item * C<directories> - directories to check recursively, supports string or arrayref
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: inconsistent, doesn't match the naming in the code and misleading because it looks like these are named parameters...

```suggestion
=item * C<$path> - path or arrayref of directories to check (recursively)
```

File: lib/Test/Pod/CoverageChange.pm [hunk 40]
```diff
@@ -0,0 +1,246 @@
...
+
+=head2 check_pod_coverage
+
+Checks POD coverage for all the modules that exists under a given directory.
+Passes the $allowed_naked_packages to the L<Test::Pod::CoverageChange::check_allowed_naked_packages>
+Ignores the packages in the C<$ignored_packages> parameter
+
+=over 4
+
+=item C<directories> - Directories to check recursively, supports string or arrayref
```
Comment (tom-binary) [inline lib/Test/Pod/CoverageChange.pm]: same issue here.

File: t/00-basic.t [hunk 41]
```diff
@@ -0,0 +1,11 @@
+use strict;
```
Comment (tom-binary) [inline t/00-basic.t]: this file can be deleted, it's already covered by dzil

File: lib/Test/Pod/CoverageChange.pm (hunk 32 above)
Comment (ramin-fs) [reply lib/Test/Pod/CoverageChange.pm]: IMHO `check` functionality can be expressed by the package namespace itself. But if your concern is about the conflict problem then I can change the name.

File: dist.ini (hunk 25 above)
Comment (ramin-fs) [reply dist.ini]: Required plugin Dist::Zilla::Plugin::Author::DERIV isn't installed.

Run 'dzil authordeps' to see a list of all required plugins.
You can pipe the list to your CPAN client to install or update them:

    dzil authordeps --missing | cpanm

I noticed that there is a difference between the required module and the module that dzil has tried to install:
// dzil is trying to install this plugin after I use [Author::DERIV]
Dist::Zilla::Plugin::Author::DERIV

// we can install our plugin using this namespace
Dist::Zilla::PluginBundle::Author::DERIV

Comment (zakame) [reply dist.ini]: I think its supposed to be

```
[@Author::DERIV]
```

precisely because it is a PluginBundle, not just a single Plugin.

See also https://metacpan.org/pod/Dist::Zilla::Starter

File: lib/Test/Pod/CoverageChange.pm (hunk 32 above)
Comment (ramin-fs) [reply lib/Test/Pod/CoverageChange.pm]: BTW your suggested name already in use in this module: https://metacpan.org/pod/Test::Pod::Coverage#pod_coverage_ok(-$module,-%5B$parms,-%5D-$msg-) and we are using it in our module as well.

Comment (ramin-fs) [reply lib/Test/Pod/CoverageChange.pm]: I've changed it to `pod_coverage_syntax_ok`

**Comment by ramin-fs** [comment]
This PR was closed because it replaced with [this PR](https://github.com/binary-com/perl-Test-Pod-CoverageChange/pull/2)
